### ISSUE TITLE
search feature is strange

### ISSUE THREAD

**Issue by ddauto** [PHPSocialNetwork/phpfastcache#288]
Hi, 
Thanks for this great class.

I encounter a strange search problem.
I patched the DriverAbstract.php with the newest code and changed below, cause I didn't install sqlite:
protected function _tagCaching()
{
    return CacheManager::Memcached(
      array(
        //"path" => $this->config[ 'path' ],
        "cache_method" => 2,
      )
    );
}

I use a class with a static function.

public static function search($keyword, $search_in_value=false)
{
    $cache = CacheManager::Memcached();
    
    echo $cache->isExisting($keyword); //**This will echo 1
    
    $rtn=$cache->search($keyword, $search_in_value);  //**This will get array(0) {}
    if (is_array($rtn)){
        return null;
    }
    else{
        return key($rtn);
    }
}

But if I set the cache in this function before search, then the search can get the result.

Thanks for your help.

**Comment by khoaofgod** [comment]
Have you followed this setup:
https://github.com/PHPSocialNetwork/phpfastcache/blob/final/examples/Search.YourCaching.php

"allow_search"  =>  true

**Comment by ddauto** [comment]
Hi,
Thanks for quick response.
Yes, I enabled the option in Core/phpFastCache.php public static $config 
and I changed below code also, I can't remove both of "path" and "cache_method", it will broken the whole php.

but it still not works.

    protected function _tagCaching()
    {
        return CacheManager::Memcached(
          array(
            "path" => $this->config[ 'path' ]
            //"cache_method" => 2,
          )
        );
    }

**Comment by ddauto** [comment]
I seems understand now, is the search function only can search the cache content that "set" in the same one request?
For example, if a.php set the key="abc", value="def", then b.php search the key "abc", it can't be searched, right? even the cache stored in memcached.

Cause If I set the key="abc" in a.php and search this key following the set, it can search the key correctly.

thanks.

**Comment by khoaofgod** [comment]
just remove the whole array(
        "path" => $this->config[ 'path' ]
        //"cache_method" => 2,
      ) 

The search function can use anytime, not in same request. However if your memcached restart, the indexing will be clean out.

I have a solution for you, use ::Files with cache_method" => 2 instead of Memcached.
It will works.

**Comment by ddauto** [comment]
Hi, all my environment is sample.

[a.php]

<?php
require_once $_SERVER['DOCUMENT_ROOT'] . '/include/config_inc.php';

use phpFastCache\CacheManager;

$key="test";
$cache = CacheManager::Memcached();
if (trim($_REQUEST['s'])=="Y"){
	$cache->delete($key);
	$cache->set($key, "value1", 10000);
}
   	
echo "isExist=" . $cache->isExisting($key) . "<BR>";

$rtn=$cache->search($key, false);
var_dump($rtn);
?>

[DriverAbstract.php]
protected function _tagCaching()
{
	return CacheManager::Memcached(
        array('127.0.0.1', 11211, 1)
    );
}

Note: if I used return CacheManager::Memcached();  ==> the whole package and php won't work.

[Test]
When I accessed:
a.php?s=Y  => $rtn has correct search result.

soon I accessed:
a.php      => $rtn didn't search the result.


Cause I used server farm so I used a shared memcached server as cache storage, so I am not sure if I use CacheManager::files in _tagCacheing() is suitable.

Thanks again.

**Comment by Geolim4** [comment]
Hi !
Did you tried with  getInstance('Memcached') instead ?

**Comment by ddauto** [comment]
Hi,

<?php
require_once $_SERVER['DOCUMENT_ROOT'] . '/include/config_inc.php';

use phpFastCache\CacheManager;

$cache = CacheManager::getInstance();
    
$key="test";
//$cache = CacheManager::Memcached();
$cache = CacheManager::getInstance();
if (trim($_REQUEST['s'])=="Y"){
	$cache->delete($key);
	$cache->set($key, "value1", 10000);
}
   	
echo "isExist=" . $cache->isExisting($key) . "<BR>";
echo "value=" . $cache->get($key) . "<BR>";

$rtn=$cache->search($key, true);

var_dump($rtn);
?>

Unfortunately, the result is the same.

Thanks.

**Comment by Geolim4** [comment]
> $cache = CacheManager::getInstance();
is wrong, try $cache = CacheManager::getInstance('Memcached');

:)

**Comment by ddauto** [comment]
<?php
require_once $_SERVER['DOCUMENT_ROOT'] . '/include/config_inc.php';

use phpFastCache\CacheManager;

//$cache = CacheManager::getInstance();
    
$key="test";
//$cache = CacheManager::Memcached();
$cache = CacheManager::getInstance('Memcached');
if (trim($_REQUEST['s'])=="Y"){
	$cache->delete($key);
	$cache->set($key, "value1", 10000);
}
   	
echo "isExist=" . $cache->isExisting($key) . "<BR>";
echo "value=" . $cache->get($key) . "<BR>";

$rtn=$cache->search($key, true);
var_dump($rtn);
?>

Unfortunately, the result is the same. ?s=Y then search has result, soon, request without ?s=Y then isExist=1
value=value1
array(0) { }

If it still not works, it's ok, I will give up the search method and use other workaround to solve.

Thanks.

**Comment by Geolim4** [comment]
Can you use the tags instead ?

**Comment by ddauto** [comment]
<?php
error_reporting(-1);
require_once $_SERVER['DOCUMENT_ROOT'] . '/include/config_inc.php';

use phpFastCache\CacheManager;

$key="test";
$cache = CacheManager::getInstance('Memcached');
if (trim($_REQUEST['s'])=="Y"){

    $cache->set($key, "value1", 10000, array("tags" => array("tag1","tag2","tag3")));
    
}
   	
$list = $cache->getTags(array("tag1"));
var_dump($list);

?>

Still not work.
test.php?s=Y => array(1) { ["tag1"]=> array(1) { ["test"]=> string(6) "value1" } }
test.php     => array(1) { ["tag1"]=> array(0) { } }

But I found that has warning message like:

Notice: Undefined index: memcache in /var/www/html_2016/vendor/phpfastcache/phpfastcache/src/phpFastCache/Drivers/memcached.php on line 74

Notice: Undefined index: memcache in /var/www/html_2016/vendor/phpfastcache/phpfastcache/src/phpFastCache/Drivers/memcached.php on line 74

Notice: Undefined index: memcache in /var/www/html_2016/vendor/phpfastcache/phpfastcache/src/phpFastCache/Drivers/memcached.php on line 74

I am not sure if it's the root cause?

Thanks.

**Comment by khoaofgod** [comment]
Just change the sqlite to Memcached. Then run examples/search and tags. To
see if the results you get is good or not.

**Comment by ddauto** [comment]
Hi, 
I have time to trace in the codes and I seems found the root cause.
The root cause is in:
DriverAbstract.php     "getTags" function
When I marked below line, and search feature and tag feature is ok, I found this issue is because when I use "search" or "getTags" method second time, it will can't get the result anymore.

$this->_tagCaching()->set($this->_getTagName($tag), $list, $tmp);

this line seems pass the absolute time to the "set" function,it seems should use "period time" instead.
Hope my understand is right.

Thanks.
